src/script/interpreter.cpp

```diff
@@ -1407,12 +1646,33 @@ bool GenericTransactionSignatureChecker<T>::CheckSig(const std::vector<unsigned
 
     uint256 sighash = SignatureHash(scriptCode, *txTo, nIn, nHashType, amount, sigversion, this->txdata);
 
-    if (!VerifySignature(vchSig, pubkey, sighash))
+    if (!VerifyECDSASignature(vchSig, pubkey, sighash))
         return false;
 
     return true;
 }
 
+template <class T>
+bool GenericTransactionSignatureChecker<T>::CheckSchnorrSignature(const std::vector<unsigned char>& sig_in, const std::vector<unsigned char>& pubkey_in, SigVersion sigversion, const ScriptExecutionData& execdata) const
+{
+    std::vector<unsigned char> sig(sig_in);
+    if (sig.empty()) return false;
+
+    if (pubkey_in.size() != 32) return false;
+    XOnlyPubKey pubkey{uint256(pubkey_in)};
```

Comment:
I'm not sure about this. BIP340 public keys are just 32-byte arrays, not abstract objects that require serialization. So there shouldn't be a concept of an "invalid public key" that's distinct from signature validation failure, and I'd prefer not to introduce such a concept to XOnlyPubKey.

Arguably, this could be an `assert(pubkey_in.size() == 32);` even. If a different length is used, it's a code error.

Reply:
Made a few changes here, including passing pubkey/signature as a Span (which means one less useless vector copy during signature validation). The size is now asserted to be 32 bytes (both in XOnlyPubKey (where it's required), and in CheckSchnorrSignature (which makes decisions based on the length).